platform/platform-crypto-provider/src/main/java/org/codice/ddf/platform/crypto/provider/MerlinExtension.java

```diff
@@ -0,0 +1,340 @@
+                    x509certs[0] = certs[0];
+                    System.arraycopy(foundCertChain, 0, x509certs, 1, foundCertChain.length);
+
+                    List<X509Certificate> certList = Arrays.asList(certs);
+                    CertPath path = getCertificateFactory().generateCertPath(certList);
+
+                    try {
+                        validator.validate(path, param);
+                        // We have a valid cert path at this point so break
+                        validatorException = null;
+                        break;
+                    } catch (java.security.cert.CertPathValidatorException e) {
+                        validatorException = e;
+                    }
+                }
+
+                if (validatorException != null) {
+                    throw validatorException;
+                }
+            } else {
+                List<X509Certificate> certList = Arrays.asList(certs);
+                CertPath path = getCertificateFactory().generateCertPath(certList);
+
+                validator.validate(path, param);
+            }
+        } catch (NoSuchProviderException | NoSuchAlgorithmException
+                | CertificateException | InvalidAlgorithmParameterException
+                | java.security.cert.CertPathValidatorException
+                | KeyStoreException e) {
+            throw new WSSecurityException(
+                    WSSecurityException.ErrorCode.FAILURE, e, "certpath"
+            );
+        }
+
+        // Finally check Cert Constraints
+        if (!matchesSubjectDnPattern(certs[0], subjectCertConstraints)) {
+            throw new WSSecurityException(WSSecurityException.ErrorCode.FAILED_AUTHENTICATION);
+        }
+    }
+
+    private void buildTrustAnchors(Set<TrustAnchor> trustAnchorSet,
+            Enumeration<String> truststoreAliases) throws KeyStoreException {
+        while (truststoreAliases.hasMoreElements()) {
+            String alias = truststoreAliases.nextElement();
+            X509Certificate cert = (X509Certificate) truststore.getCertificate(alias);
+            if (cert != null) {
+                byte[] nameConstraints = null;
+                try {
+                    nameConstraints = extractNameConstraints(cert);
+                } catch (IOException e) {
+                    LOG.debug("Unable to decode name constraints.", e);
+                }
+                TrustAnchor trustAnchor = new TrustAnchor(cert, nameConstraints);
+                trustAnchorSet.add(trustAnchor);
+            }
+        }
+    }
+
+    private static byte[] extractNameConstraints(X509Certificate x509Certificate)
+            throws IOException {
+        byte[] bytes = x509Certificate.getExtensionValue(NAME_CONSTRAINTS_OID);
+        if (bytes == null || bytes.length <= 0) {
+            return null;
+        }
+
+        switch (bytes[0]) {
+        case DERDecoder.TYPE_OCTET_STRING: {
```

Comment:
It might not be necessary to switch and we could change to always assume that it is correctly encoded as such, but this is here as a belt&suspenders safety check. While the Java docs are clear about what to expect be returned, the relevant RFCs are more ambiguous about Sequence vs. encoded Octet String. It's a cheap check to look at the first byte.